epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java

```diff
@@ -0,0 +1,180 @@
+import android.support.v7.widget.RecyclerView;
+import android.support.v7.widget.RecyclerView.LayoutManager;
+import android.support.v7.widget.RecyclerView.OnChildAttachStateChangeListener;
+import android.support.v7.widget.RecyclerView.OnScrollListener;
+import android.support.v7.widget.RecyclerView.ViewHolder;
+import android.util.SparseArray;
+import android.view.View;
+import android.view.View.OnLayoutChangeListener;
+
+/**
+ * Helper class to handle visibility event from an {@link com.airbnb.epoxy.EpoxyRecyclerView}.
+ *
+ * <p>
+ * - Visible Event: this event is triggered when at least one pixel of the Component is visible.
+ * <p>
+ * - Invisible Event: this event is triggered when the Component no longer has any pixels on the
+ * screen.
+ * <p>
+ * - Focused Visible Event: this event is triggered when either the Component occupies at least
+ * half of the viewport, or, if the Component is smaller than half the viewport, when it is fully
+ * visible.
+ * <p>
+ * - Unfocused Visible Event: this event is triggered when the Component is no longer focused, i.e.
+ * it is not fully visible and does not occupy at least half the viewport.
+ * <p>
+ * - Full Impression Visible Event: this event is triggered when the entire Component has passed
+ * through the viewport at some point.
+ * <p>
+ * - VisibilityChangedEvent: this event is triggered when the visibility of the Component on the
+ * screen changes.
+ */
+class EpoxyVisibilityTracker {
+
+  /** Maintain visibility item indexed by view id (identity hashcode) */
+  private final SparseArray<EpoxyVisibilityItem> visibilityIdToItemMap = new SparseArray<>();
+
+  /** listener use to process scroll, layout and attaches events */
+  private final Listener listener = new Listener();
+
+  @Nullable
+  private RecyclerView attachedRecyclerView = null;
+
+  @Nullable
+  private LinearLayoutManager attachedLinearLayoutManager = null;
+
+  /**
+   * Attach the tracker.
+   *
+   * @param recyclerView The EpoxyRecyclerView.
+   */
+  void attach(@NonNull EpoxyRecyclerView recyclerView) {
+    attachedRecyclerView = recyclerView;
+    recyclerView.addOnScrollListener(this.listener);
+    recyclerView.addOnLayoutChangeListener(this.listener);
+    recyclerView.addOnChildAttachStateChangeListener(this.listener);
+  }
+
+  /**
+   * Detach the tracker
+   *
+   * @param recyclerView The EpoxyRecyclerView.
+   */
+  void detach(@NonNull EpoxyRecyclerView recyclerView) {
+    recyclerView.removeOnScrollListener(this.listener);
+    recyclerView.removeOnLayoutChangeListener(this.listener);
+    recyclerView.removeOnChildAttachStateChangeListener(this.listener);
+    attachedRecyclerView = null;
+    attachedLinearLayoutManager = null;
+  }
+
+  @Nullable
+  private LinearLayoutManager getLinearLayoutManager() {
+    if (attachedLinearLayoutManager == null && attachedRecyclerView != null) {
+      LayoutManager lm = attachedRecyclerView.getLayoutManager();
+      if (lm instanceof LinearLayoutManager) {
```

Comment:
maybe a log to say this doesn't support anything other than LinearLayoutManager

Reply:
I would crash if they enable it and it's not a LinearLayoutManager with an IllegalStateException and helpful error message.